pkg/core/workflow_store.go

```diff
@@ -62,44 +62,85 @@ func (ws *WorkflowStore) AddWorkflow(ctx context.Context, w *workflow.Workflow)
+	result = make(map[string][]*domain.AssignedCodeset, len(ws.items))
 	if workflowName != nil {
-		if _, exists := ws.items[*workflowName]; exists {
-			result[*workflowName] = ws.items[*workflowName].assignedCodesets
+		if sw, exists := ws.items[*workflowName]; exists && len(sw.assignedCodesets) > 0 {
+			result[*workflowName] = sw.assignedCodesets
 		}
 		return
 	}
 	for _, sw := range ws.items {
-		result[sw.workflow.Name] = sw.assignedCodesets
+		if len(sw.assignedCodesets) > 0 {
+			result[sw.workflow.Name] = sw.assignedCodesets
+		}
 	}
 	return
 }
 
 // AddCodesetAssignment adds a codeset to the list of assigned codesets of a workflow if it does not already exists
-func (ws *WorkflowStore) AddCodesetAssignment(ctx context.Context, workflowName string, codeset *domain.Codeset) []*domain.Codeset {
+func (ws *WorkflowStore) AddCodesetAssignment(ctx context.Context, workflowName string,
+	assignedCodeset *domain.AssignedCodeset) []*domain.AssignedCodeset {
 	assignedCodesets := ws.items[workflowName].assignedCodesets
-	if !containsCodeset(assignedCodesets, codeset) {
-		assignedCodesets = append(assignedCodesets, codeset)
+	if assigned, _ := getAssignedCodeset(assignedCodesets, assignedCodeset.Codeset); assigned == nil {
+		assignedCodesets = append(assignedCodesets, assignedCodeset)
 		ws.items[workflowName].assignedCodesets = assignedCodesets
 	}
 	return assignedCodesets
 }
 
-func containsCodeset(slice []*domain.Codeset, codeset *domain.Codeset) bool {
-	for _, c := range slice {
-		if c.Project == codeset.Project && c.Name == codeset.Name {
-			return true
+// DeleteCodesetAssignment deletes a codeset from the list of assigned codesets of a workflow if it exists
+func (ws *WorkflowStore) DeleteCodesetAssignment(ctx context.Context, workflowName string, codeset *domain.Codeset) []*domain.AssignedCodeset {
+	assignedCodesets := ws.items[workflowName].assignedCodesets
+	if _, i := getAssignedCodeset(assignedCodesets, codeset); i != -1 {
+		assignedCodesets = removeAssignedCodeset(assignedCodesets, i)
+		ws.items[workflowName].assignedCodesets = assignedCodesets
+	}
+	return assignedCodesets
+}
+
+// GetAssignedCodeset returns a AssignedCodeset for the Workflow and Codeset
+func (ws *WorkflowStore) GetAssignedCodeset(ctx context.Context, workflowName string, codeset *domain.Codeset) *domain.AssignedCodeset {
+	sw, exists := ws.items[workflowName]
+	if !exists {
+		return nil
+	}
+	ac, _ := getAssignedCodeset(sw.assignedCodesets, codeset)
+	return ac
+}
+
+func getAssignedCodeset(slice []*domain.AssignedCodeset, codeset *domain.Codeset) (*domain.AssignedCodeset, int) {
+	for i, ac := range slice {
+		if ac.Codeset.Project == codeset.Project && ac.Codeset.Name == codeset.Name {
+			return ac, i
 		}
 	}
-	return false
+	return nil, -1
+}
+
+func removeAssignedCodeset(codesets []*domain.AssignedCodeset, index int) []*domain.AssignedCodeset {
+	codesets[index] = codesets[len(codesets)-1]
+	return codesets[:len(codesets)-1]
```

Comment:
This is smart; however I still have a question, wouldn't it be better to use map instead of array?

Reply:
How so?

Reply:
Because map is better suited for searching for an item, assuming you index it correctly?

Reply:
I still do not understand what you are suggesting. Use a map instead of a list for `assignedCodesets`? What would I use as key?

Reply:
Exactly.

I assume you'd need to use nested map, first index by project name and then by codeset name. Which is a bit of a bummer, I admit.

Reply:
So it's probably not worth the effort.

Reply:
It took me a while to understand what you guys are talking about here, but now that I'm all caught up, I have to agree with @jsuchome , this would have been better modeled as a map. You can use a golang struct as a key. In fact, this exact case is covered here quite nicely: https://blog.golang.org/maps#TOC_5. 

I find it a bit strange that you would go to those great lengths of converting an array into a map to reduce lookup time for an array that wouldn't normally exceed a couple of items, but here (in `getAssignedCodeset`) you have no problem iterating through the entire array to find what you're looking for. Anyway, it's a bit too late now, this is something I should've recommended in the `codeset assign` implementation, and this PR is already reaching sizes that make code-review a burden, so we should postpone it for a followup. My apologies for not catching this earlier.

The other way to look at this is that we need to revisit the internal representation of these inter-object relationships anyway when we choose a persistent backend for the state stored by FuseML. If we choose a relational database, for example, a lot of operations () will be implemented as SQL queries, so we won't need to reinvent the wheel here. We'll have another chance then to get this right, and to account for performance and scalability.

Reply:
BTW although I think maps might make more sense from the view of code purity, I do not think it _needs_ to be done